File: server/src/workspace-engine.js
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { getPlan, evaluate, describe } from './quota.js';

const fsp = fs.promises;

export const MAX_READ_BYTES = 1024 * 1024; // 1 MB shown in the viewer; more via ?raw=1
export const MAX_EXEC_OUTPUT = 64 * 1024;

export function nowIso() {
  return new Date().toISOString();
}

export function shortId(bytes = 6) {
  return crypto.randomBytes(bytes).toString('hex');
}

/** Deterministic content hash used by the sync engine for delta detection. */
export function sha256(buf) {
  return crypto.createHash('sha256').update(buf).digest('hex');
}

/**
 * On-disk layout, rooted at dataRoot:
 *
 *   <dataRoot>/
 *     sessions/<sessionId>/files/     <- the workspace the agent edits
 *     sessions/<sessionId>/meta.json
 *     sessions/<sessionId>/oplog.jsonl
 *     sessions/<sessionId>/snapshots/
 *     sync/remote/<sessionId>/...     <- the local stand-in for cloud storage
 */
export class WorkspaceEngine {
  constructor({ dataRoot, planId = 'unlimited' }) {
    this.dataRoot = path.resolve(dataRoot);
    this.plan = getPlan(planId);
    fs.mkdirSync(path.join(this.dataRoot, 'sessions'), { recursive: true });
    fs.mkdirSync(path.join(this.dataRoot, 'sync', 'remote'), { recursive: true });
  }

  sessionDir(sessionId) {
    return path.join(this.dataRoot, 'sessions', sessionId);
  }

  rootFor(sessionId) {
    return path.join(this.sessionDir(sessionId), 'files');
  }

  /**
   * Resolve an untrusted relative path inside a session workspace.
   * Rejects traversal instead of silently clamping, so a confused agent gets a
   * clear error rather than writing outside its sandbox.
   */
  resolve(sessionId, relPath = '.') {
    const root = this.rootFor(sessionId);
    const clean = path.posix.normalize(String(relPath).replace(/\\/g, '/')).replace(/^\/+/, '');
    const abs = path.resolve(root, clean);
    if (abs !== root && !abs.startsWith(root + path.sep)) {
      const err = new Error(`Path keluar dari workspace: ${relPath}`);
      err.code = 'EPATHOUTSIDE';
      throw err;
    }
    return abs;
  }

  // ---------------------------------------------------------------- lifecycle

  async createSession({ title = 'Workspace baru', model = 'arena-agent-1', planId } = {}) {
    const id = `ws_${nowIso().slice(0, 10).replace(/-/g, '')}_${shortId(4)}`;
    const dir = this.sessionDir(id);
    await fsp.mkdir(path.join(dir, 'files'), { recursive: true });
    await fsp.mkdir(path.join(dir, 'snapshots'), { recursive: true });
    const meta = {
      id,
      title,
      model,
      plan: planId || this.plan.id,
      createdAt: nowIso(),
      updatedAt: nowIso(),
      messageCount: 0,
      status: 'idle',
    };
    await fsp.writeFile(path.join(dir, 'meta.json'), JSON.stringify(meta, null, 2));
    await this.writeFile(id, 'README.md', seedReadme(title));
    await this.appendOp(id, { type: 'session.create', path: null, detail: { title } });
    return meta;
  }

  async readMeta(sessionId) {
    const file = path.join(this.sessionDir(sessionId), 'meta.json');
    if (!fs.existsSync(file)) {
      const err = new Error(`Workspace ${sessionId} tidak ditemukan`);
      err.code = 'ENOSESSION';
      throw err;
    }
    return JSON.parse(await fsp.readFile(file, 'utf8'));
  }

  async writeMeta(sessionId, patch) {
    const meta = { ...(await this.readMeta(sessionId)), ...patch, updatedAt: nowIso() };
    await fsp.writeFile(path.join(this.sessionDir(sessionId), 'meta.json'), JSON.stringify(meta, null, 2));
    return meta;
  }

  async listSessions() {
    const base = path.join(this.dataRoot, 'sessions');
    const entries = await fsp.readdir(base, { withFileTypes: true }).catch(() => []);
    const out = [];
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      try {
        const meta = await this.readMeta(entry.name);
        out.push(meta);
      } catch {
        /* half-written workspace, skip */
      }
    }
    out.sort((a, b) => (a.updatedAt < b.updatedAt ? 1 : -1));
    return out;
  }

  /**
   * "Unlimited" also means unlimited clones: forking a workspace copies the
   * files and rewrites the oplog cursor so the copy syncs independently.
   */
  async forkSession(sessionId, { title } = {}) {
    const src = await this.readMeta(sessionId);
    const copy = await this.createSession({
      title: title || `${src.title} (copy)`,
      model: src.model,
    });
    await copyDir(this.rootFor(sessionId), this.rootFor(copy.id));
    await this.appendOp(copy.id, { type: 'session.fork', path: null, detail: { from: sessionId } });
    return copy;
  }

  async deleteSession(sessionId) {
    await this.readMeta(sessionId); // throws when unknown
    await fsp.rm(this.sessionDir(sessionId), { recursive: true, force: true });
    return { id: sessionId, deleted: true };
  }

  // -------------------------------------------------------------------- files

  async listFiles(sessionId, relPath = '.') {
    const abs = this.resolve(sessionId, relPath);
    const entries = await fsp.readdir(abs, { withFileTypes: true });
    // Normalise the listing root so entries always carry a workspace-relative
    // POSIX path ("src/index.js"), never a leading slash — clients paste these
    // straight back into /files/<path> and into the chat tools.
    const relRoot = path.posix.normalize(String(relPath).replace(/\\/g, '/')).replace(/^\/+/, '');
    const prefix = relRoot === '.' || relRoot === '' ? '' : relRoot;
    const items = await Promise.all(
      entries.map(async (entry) => {
        const full = path.join(abs, entry.name);
        const st = await fsp.stat(full).catch(() => null);
        return {
          name: entry.name,
          path: prefix ? `${prefix}/${entry.name}` : entry.name,
          type: entry.isDirectory() ? 'dir' : 'file',
          size: st ? st.size : 0,
          modifiedAt: st ? st.mtime.toISOString() : null,
        };
      }),
    );
    items.sort((a, b) => (a.type === b.type ? a.name.localeCompare(b.name) : a.type === 'dir' ? -1 : 1));
    return items;
  }

  async readFile(sessionId, relPath) {
    const abs = this.resolve(sessionId, relPath);
    const st = await fsp.stat(abs);
    if (st.isDirectory()) {
      const err = new Error(`${relPath} adalah direktori`);
      err.code = 'EISDIR';
      throw err;
    }
    const buf = await fsp.readFile(abs);
    return {
      path: relPath,
      size: buf.length,
      truncated: buf.length > MAX_READ_BYTES,
      base64: buf.toString('base64'),
      text: isProbablyText(buf) ? buf.subarray(0, MAX_READ_BYTES).toString('utf8') : null,
      hash: sha256(buf),
      modifiedAt: st.mtime.toISOString(),
    };
  }

  /**
   * The single write path. Everything (agent tool, REST API, sync pull) goes
   * through here so the oplog + usage accounting can never drift.
   */
  async writeFile(sessionId, relPath, content, { mode = 0o644, source = 'api' } = {}) {
    const usage = await this.usage(sessionId);
    const incoming = Buffer.byteLength(content);
    const decision = evaluate(this.plan, usage, { incomingBytes: incoming });
    if (!decision.ok) {
      const err = new Error('Kuota terlampaui');
      err.code = 'EQUOTA';
      throw err;
    }

    const abs = this.resolve(sessionId, relPath);
    const previous = fs.existsSync(abs) ? await fsp.readFile(abs).catch(() => null) : null;
    await fsp.mkdir(path.dirname(abs), { recursive: true });
    await fsp.writeFile(abs, content, { mode });

    const stat = await fsp.stat(abs);
    const entry = {
      type: 'file.write',
      path: relPath,
      source,
      at: nowIso(),
      detail: {
        bytes: incoming,
        hash: sha256(Buffer.from(content)),
        previousHash: previous ? sha256(previous) : null,
        previousSize: previous ? previous.length : 0,
        created: previous === null,
      },
    };
    await this.appendOp(sessionId, entry);
    await this.writeMeta(sessionId, {});
    return { path: relPath, bytes: incoming, created: previous === null, size: stat.size, warnings: decision.warnings };
  }

  async deletePath(sessionId, relPath, { source = 'api' } = {}) {
    const abs = this.resolve(sessionId, relPath);
    if (abs === this.rootFor(sessionId)) {
      const err = new Error('Tidak bisa menghapus akar workspace');
      err.code = 'EROOT';
      throw err;
    }
    const existed = fs.existsSync(abs);
    await fsp.rm(abs, { recursive: true, force: true });
    await this.appendOp(sessionId, { type: 'file.delete', path: relPath, source, at: nowIso(), detail: { existed } });
    return { path: relPath, deleted: existed };
  }

  async mkdir(sessionId, relPath, { source = 'api' } = {}) {
    const abs = this.resolve(sessionId, relPath);
    await fsp.mkdir(abs, { recursive: true });
    await this.appendOp(sessionId, { type: 'dir.create', path: relPath, source, at: nowIso(), detail: {} });
    return { path: relPath, created: true };
  }

  async movePath(sessionId, from, to, { source = 'api' } = {}) {
    const src = this.resolve(sessionId, from);
    const dst = this.resolve(sessionId, to);
    await fsp.mkdir(path.dirname(dst), { recursive: true });
    await fsp.rename(src, dst);
    await this.appendOp(sessionId, { type: 'file.move', path: to, source, at: nowIso(), detail: { from } });
    return { from, to };
  }

  /** Copy a whole tree — used by snapshots and by forkSession. */
  async snapshot(sessionId, label = `snap_${shortId(3)}`) {
    const dest = path.join(this.sessionDir(sessionId), 'snapshots', label);
    await fsp.mkdir(dest, { recursive: true });
    await copyDir(this.rootFor(sessionId), dest);
    const files = await walk(dest);
    await this.appendOp(sessionId, {
      type: 'snapshot.create',
      path: null,
      source: 'system',
      at: nowIso(),
      detail: { label, files: files.length },
    });
    return { label, files: files.length };
  }

  // -------------------------------------------------------------- usage / ops

  /**
   * Measured, never limited. Returns Infinity for the plan caps so the UI can
   * print "∞" without special casing.
   */
  async usage(sessionId) {
    // readMeta first so an unknown/deleted workspace is a 404, not a silent
    // "0 files" answer that looks like a real empty workspace.
    const meta = await this.readMeta(sessionId);
    const root = this.rootFor(sessionId);
    const files = await walk(root);
    let totalBytes = 0;
    let largest = 0;
    for (const f of files) totalBytes += f.size;
    for (const f of files) largest = Math.max(largest, f.size);
    return {
      sessionId,
      files: files.length,
      directories: await countDirs(root),
      totalBytes,
      largestFileBytes: largest,
      messageCount: meta.messageCount || 0,
      plan: describe(this.plan),
    };
  }

  async globalUsage() {
    const sessions = await this.listSessions();
    let totalBytes = 0;
    let files = 0;
    for (const s of sessions) {
      const u = await this.usage(s.id);
      totalBytes += u.totalBytes;
      files += u.files;
    }
    return {
      workspaces: sessions.length,
      files,
      totalBytes,
      plan: describe(this.plan),
    };
  }

  async appendOp(sessionId, op) {
    const record = { seq: await this.nextSeq(sessionId), ...op, at: op.at || nowIso() };
    const file = path.join(this.sessionDir(sessionId), 'oplog.jsonl');
    await fsp.appendFile(file, JSON.stringify(record) + '\n');
    return record;
  }

  async nextSeq(sessionId) {
    const file = path.join(this.sessionDir(sessionId), 'seq');
    let n = 0;
    try {
      n = parseInt(await fsp.readFile(file, 'utf8'), 10) || 0;
    } catch {
      n = 0;
    }
    n += 1;
    await fsp.writeFile(file, String(n));
    return n;
  }

  /**
   * Delta feed for the sync engine and the web client's activity panel.
   * `cursor` is the last seq the caller has seen; 0 means "everything".
   */
  async opsSince(sessionId, cursor = 0) {
    const file = path.join(this.sessionDir(sessionId), 'oplog.jsonl');
    const raw = await fsp.readFile(file, 'utf8').catch(() => '');
    const ops = raw
      .split('\n')
      .filter(Boolean)
      .map((line) => {
        try {
          return JSON.parse(line);
        } catch {
          return null;
        }
      })
      .filter((o) => o && o.seq > cursor);
    return { cursor: ops.length ? ops[ops.length - 1].seq : cursor, ops };
  }

  /** Materialised file list with hashes — what the sync engine diffs against. */
  async manifest(sessionId) {
    const root = this.rootFor(sessionId);
    const files = await walk(root);
    const out = [];
    for (const f of files) {
      const rel = path.relative(root, f.path).split(path.sep).join('/');
      const buf = await fsp.readFile(f.path);
      out.push({ path: rel, size: f.size, hash: sha256(buf), modifiedAt: f.mtime });
    }
    out.sort((a, b) => a.path.localeCompare(b.path));
    return { sessionId, generatedAt: nowIso(), files: out };
  }
}

// ------------------------------------------------------------------- helpers

export async function walk(dir) {
  const out = [];
  const stack = [dir];
  while (stack.length) {
    const current = stack.pop();
    const entries = await fsp.readdir(current, { withFileTypes: true }).catch(() => []);
    for (const e of entries) {
      const full = path.join(current, e.name);
      if (e.isDirectory()) stack.push(full);
      else if (e.isFile()) {
        const st = await fsp.stat(full).catch(() => null);
        if (st) out.push({ path: full, size: st.size, mtime: st.mtime.toISOString() });
      }
    }
  }
  return out;
}

async function countDirs(dir) {
  let n = 0;
  const stack = [dir];
  while (stack.length) {
    const current = stack.pop();
    const entries = await fsp.readdir(current, { withFileTypes: true }).catch(() => []);
    for (const e of entries) {
      if (e.isDirectory()) {
        n += 1;
        stack.push(path.join(current, e.name));
      }
    }
  }
  return n;
}

async function copyDir(src, dest) {
  await fsp.mkdir(dest, { recursive: true });
  const entries = await fsp.readdir(src, { withFileTypes: true }).catch(() => []);
  for (const e of entries) {
    const s = path.join(src, e.name);
    const d = path.join(dest, e.name);
    if (e.isDirectory()) await copyDir(s, d);
    else await fsp.copyFile(s, d);
  }
}

export function isProbablyText(buf) {
  const sample = buf.subarray(0, 512);
  if (sample.length === 0) return true;
  let suspicious = 0;
  for (const byte of sample) if (byte === 0) suspicious += 1;
  return suspicious === 0;
}

export function seedReadme(title) {
  return `# ${title}

Workspace ini **tidak punya kuota**: simpan file sebesar apa pun, buat session sebanyak apa pun.

- \`files/\` di server = isi workspace ini
- Setiap perubahan dicatat di \`oplog.jsonl\` dan bisa di-sinkronkan
- Agent punya tools: read_file, write_file, edit_file, list_files, delete_file, exec

Coba minta agent: "buatkan script python yang menghitung fibonacci lalu jalankan".
`;
}

File: android/app/src/core/kotlin/dev/arenalite/core/workspace/WorkspaceEngine.kt
package dev.arenalite.core.workspace

import dev.arenalite.core.json.parseJson
import dev.arenalite.core.json.toJsonValue
import dev.arenalite.core.model.FileManifest
import dev.arenalite.core.model.FileManifestEntry
import dev.arenalite.core.model.OpLogEntry
import dev.arenalite.core.model.QuotaPolicy
import dev.arenalite.core.model.WorkspaceEntry
import dev.arenalite.core.model.WorkspaceSession
import dev.arenalite.core.model.WorkspaceUsage
import dev.arenalite.core.util.FileSystem
import dev.arenalite.core.util.NoSuchWorkspaceException
import dev.arenalite.core.util.PathEscapeException
import dev.arenalite.core.util.WorkspaceFileException
import dev.arenalite.core.util.isProbablyText
import dev.arenalite.core.util.normalizeRelative
import dev.arenalite.core.util.nowIso
import dev.arenalite.core.util.randomHex
import dev.arenalite.core.util.sha256Hex

/**
 * On-device workspace engine — the Android twin of the server's engine.
 *
 * Layout under `fs.root()`:
 *   sessions/<id>/files/       the workspace the agent edits
 *   sessions/<id>/meta.json
 *   sessions/<id>/oplog.jsonl
 *   sessions/<id>/seq
 *   sessions/<id>/snapshots/
 *
 * There is deliberately no size or count check anywhere in here: [QuotaPolicy]
 * is consulted only to produce advisory warnings for the UI.
 */
class WorkspaceEngine(
    private val fs: FileSystem,
    private val policy: QuotaPolicy = QuotaPolicy.UNLIMITED,
) {
    private val base: String get() = fs.join(fs.root(), "sessions")

    fun sessionDir(sessionId: String): String = fs.join(base, sessionId)
    fun rootFor(sessionId: String): String = fs.join(sessionDir(sessionId), "files")

    /** Resolve an untrusted relative path inside a session, rejecting traversal. */
    fun resolve(sessionId: String, relative: String): String {
        val clean = normalizeRelative(relative)
        return fs.join(rootFor(sessionId), clean)
    }

    // ------------------------------------------------------------- lifecycle

    fun createSession(title: String = "Workspace baru", model: String = "arena-agent-1"): WorkspaceSession {
        fs.mkdirs(base)
        val stamp = nowIso().take(10).replace("-", "")
        val id = "ws_${stamp}_${randomHex(4)}"
        fs.mkdirs(rootFor(id))
        fs.mkdirs(fs.join(sessionDir(id), "snapshots"))
        val session = WorkspaceSession(
            id = id,
            title = title,
            model = model,
            plan = policy.id,
            createdAt = nowIso(),
            updatedAt = nowIso(),
        )
        writeMeta(session)
        writeFile(id, "README.md", seedReadme(title))
        appendOp(id, OpLogEntry(0, "session.create", null, "system", nowIso(), mapOf("title" to title)))
        return session
    }

    fun readMeta(sessionId: String): WorkspaceSession {
        val file = fs.join(sessionDir(sessionId), "meta.json")
        if (!fs.exists(file)) throw NoSuchWorkspaceException(sessionId)
        return WorkspaceSession.fromJson(parseJson(fs.readText(file)))
    }

    fun writeMeta(session: WorkspaceSession): WorkspaceSession {
        val updated = session.copy(updatedAt = nowIso())
        fs.mkdirs(sessionDir(session.id))
        fs.writeText(fs.join(sessionDir(session.id), "meta.json"), updated.toJson().toString())
        return updated
    }

    fun updateMeta(sessionId: String, transform: (WorkspaceSession) -> WorkspaceSession): WorkspaceSession =
        writeMeta(transform(readMeta(sessionId)))

    fun listSessions(): List<WorkspaceSession> {
        if (!fs.exists(base)) return emptyList()
        return fs.list(base)
            .filter { it.isDirectory }
            .mapNotNull { runCatching { readMeta(it.name) }.getOrNull() }
            .sortedByDescending { it.updatedAt }
    }

    /** Unlimited also means unlimited clones: a fork is a full independent copy. */
    fun forkSession(sessionId: String, title: String? = null): WorkspaceSession {
        val source = readMeta(sessionId)
        val copy = createSession(title ?: "${source.title} (copy)", source.model)
        copyTree(rootFor(sessionId), rootFor(copy.id))
        appendOp(copy.id, OpLogEntry(0, "session.fork", null, "system", nowIso(), mapOf("from" to sessionId)))
        return copy
    }

    fun deleteSession(sessionId: String) {
        readMeta(sessionId) // throws when unknown
        fs.delete(sessionDir(sessionId))
    }

    // ----------------------------------------------------------------- files

    fun listFiles(sessionId: String, relative: String = "."): List<WorkspaceEntry> {
        val abs = resolve(sessionId, relative)
        if (!fs.exists(abs)) return emptyList()
        val prefix = normalizeRelative(relative)
        return fs.list(abs)
            .map { entry ->
                WorkspaceEntry(
                    name = entry.name,
                    path = if (prefix.isEmpty()) entry.name else "$prefix/${entry.name}",
                    isDirectory = entry.isDirectory,
                    size = entry.size,
                    modifiedAt = entry.modifiedAt,
                )
            }
            .sortedWith(compareBy({ !it.isDirectory }, { it.name }))
    }

    data class FileContent(
        val path: String,
        val size: Long,
        val text: String?,
        val truncated: Boolean,
        val hash: String,
        val modifiedAt: Long,
    )

    fun readFile(sessionId: String, relative: String, maxBytes: Int = MAX_READ_BYTES): FileContent {
        val abs = resolve(sessionId, relative)
        if (fs.isDirectory(abs)) throw WorkspaceFileException("$relative adalah direktori", "EISDIR")
        if (!fs.exists(abs)) throw WorkspaceFileException("$relative tidak ditemukan", "ENOENT")
        val bytes = fs.readBytes(abs)
        return FileContent(
            path = normalizeRelative(relative),
            size = bytes.size.toLong(),
            text = if (isProbablyText(bytes)) bytes.copyOf(minOf(bytes.size, maxBytes)).decodeToString() else null,
            truncated = bytes.size > maxBytes,
            hash = sha256Hex(bytes),
            modifiedAt = fs.modifiedAt(abs),
        )
    }

    /**
     * The single write path: agent tool, UI editor and sync pull all land here so
     * the oplog and usage accounting cannot drift.
     */
    fun writeFile(
        sessionId: String,
        relative: String,
        content: String,
        source: String = "api",
    ): WriteResult {
        val usage = usage(sessionId)
        val incoming = content.encodeToByteArray().size.toLong()
        val decision = policy.evaluate(usage.totalBytes, incoming)
        if (!decision.ok) throw WorkspaceFileException("Kuota terlampaui", "EQUOTA")

        val abs = resolve(sessionId, relative)
        val created = !fs.exists(abs)
        val previousHash = if (created) null else sha256Hex(fs.readBytes(abs))
        fs.mkdirs(fs.join(fs.root(), parentOf(abs)))
        fs.writeText(abs, content)

        appendOp(
            sessionId,
            OpLogEntry(
                seq = 0,
                type = "file.write",
                path = normalizeRelative(relative),
                source = source,
                at = nowIso(),
                detail = mapOf(
                    "bytes" to incoming,
                    "hash" to sha256Hex(content),
                    "previousHash" to previousHash,
                    "created" to created,
                ),
            ),
        )
        touch(sessionId)
        return WriteResult(normalizeRelative(relative), incoming, created, fs.size(abs), decision.warnings)
    }

    fun deletePath(sessionId: String, relative: String, source: String = "api"): Boolean {
        val abs = resolve(sessionId, relative)
        if (abs == rootFor(sessionId)) throw WorkspaceFileException("Tidak bisa menghapus akar workspace", "EROOT")
        val existed = fs.exists(abs)
        fs.delete(abs)
        appendOp(sessionId, OpLogEntry(0, "file.delete", normalizeRelative(relative), source, nowIso(), mapOf("existed" to existed)))
        touch(sessionId)
        return existed
    }

    fun mkdir(sessionId: String, relative: String, source: String = "api") {
        val abs = resolve(sessionId, relative)
        fs.mkdirs(abs)
        appendOp(sessionId, OpLogEntry(0, "dir.create", normalizeRelative(relative), source, nowIso()))
    }

    fun movePath(sessionId: String, from: String, to: String, source: String = "api") {
        fs.rename(resolve(sessionId, from), resolve(sessionId, to))
        appendOp(sessionId, OpLogEntry(0, "file.move", normalizeRelative(to), source, nowIso(), mapOf("from" to from)))
    }

    fun snapshot(sessionId: String, label: String = "snap_${randomHex(3)}"): SnapshotResult {
        readMeta(sessionId)
        val dest = fs.join(sessionDir(sessionId), "snapshots", label)
        fs.mkdirs(dest)
        copyTree(rootFor(sessionId), dest)
        val count = walk(dest).size
        appendOp(sessionId, OpLogEntry(0, "snapshot.create", null, "system", nowIso(), mapOf("label" to label, "files" to count)))
        return SnapshotResult(label, count)
    }

    // ----------------------------------------------------------- usage & ops

    fun usage(sessionId: String): WorkspaceUsage {
        val meta = readMeta(sessionId)
        val root = rootFor(sessionId)
        val files = walk(root)
        val dirs = countDirs(root)
        var total = 0L
        var largest = 0L
        for (f in files) {
            total += f.size
            if (f.size > largest) largest = f.size
        }
        return WorkspaceUsage(
            sessionId = sessionId,
            files = files.size,
            directories = dirs,
            totalBytes = total,
            largestFileBytes = largest,
            messageCount = meta.messageCount,
            policy = policy,
        )
    }

    fun globalUsage(): GlobalUsage {
        val sessions = listSessions()
        var files = 0
        var bytes = 0L
        for (s in sessions) {
            val u = usage(s.id)
            files += u.files
            bytes += u.totalBytes
        }
        return GlobalUsage(sessions.size, files, bytes, policy)
    }

    fun appendOp(sessionId: String, entry: OpLogEntry): OpLogEntry {
        val seq = nextSeq(sessionId)
        val record = entry.copy(seq = seq, at = if (entry.at.isEmpty()) nowIso() else entry.at)
        fs.mkdirs(sessionDir(sessionId))
        fs.appendText(fs.join(sessionDir(sessionId), "oplog.jsonl"), record.toJson().toString() + "\n")
        return record
    }

    private fun nextSeq(sessionId: String): Long {
        val file = fs.join(sessionDir(sessionId), "seq")
        val current = if (fs.exists(file)) fs.readText(file).trim().toLongOrNull() ?: 0L else 0L
        val next = current + 1
        fs.writeText(file, next.toString())
        return next
    }

    fun opsSince(sessionId: String, cursor: Long = 0): OpLogPage {
        val file = fs.join(sessionDir(sessionId), "oplog.jsonl")
        if (!fs.exists(file)) return OpLogPage(cursor, emptyList())
        val ops = fs.readText(file)
            .lineSequence()
            .filter { it.isNotBlank() }
            .mapNotNull { runCatching { OpLogEntry.fromJson(parseJson(it)) }.getOrNull() }
            .filter { it.seq > cursor }
            .toList()
        return OpLogPage(ops.lastOrNull()?.seq ?: cursor, ops)
    }

    /** Materialised file list with hashes — what the sync engine diffs against. */
    fun manifest(sessionId: String): FileManifest {
        val root = rootFor(sessionId)
        val entries = walk(root).map { file ->
            val rel = relativize(root, file.path)
            FileManifestEntry(
                path = rel,
                size = file.size,
                hash = sha256Hex(fs.readBytes(file.path)),
                modifiedAt = file.modifiedAt,
            )
        }.sortedBy { it.path }
        return FileManifest(sessionId, nowIso(), entries)
    }

    // -------------------------------------------------------------- internal

    data class WriteResult(val path: String, val bytes: Long, val created: Boolean, val size: Long, val warnings: List<String>)
    data class SnapshotResult(val label: String, val files: Int)
    data class OpLogPage(val cursor: Long, val ops: List<OpLogEntry>)
    data class GlobalUsage(val workspaces: Int, val files: Int, val totalBytes: Long, val policy: QuotaPolicy)

    private fun touch(sessionId: String) {
        runCatching { updateMeta(sessionId) { it } }
    }

    private data class WalkFile(val path: String, val size: Long, val modifiedAt: Long)

    private fun walk(dir: String): List<WalkFile> {
        if (!fs.exists(dir)) return emptyList()
        val out = mutableListOf<WalkFile>()
        val stack = ArrayDeque<String>()
        stack.addLast(dir)
        while (stack.isNotEmpty()) {
            val current = stack.removeLast()
            for (entry in fs.list(current)) {
                if (entry.isDirectory) stack.addLast(entry.path) else out.add(WalkFile(entry.path, entry.size, entry.modifiedAt))
            }
        }
        return out
    }

    private fun countDirs(dir: String): Int {
        if (!fs.exists(dir)) return 0
        var n = 0
        val stack = ArrayDeque<String>()
        stack.addLast(dir)
        while (stack.isNotEmpty()) {
            val current = stack.removeLast()
            for (entry in fs.list(current)) {
                if (entry.isDirectory) {
                    n += 1
                    stack.addLast(entry.path)
                }
            }
        }
        return n
    }

    private fun copyTree(src: String, dest: String) {
        if (!fs.exists(src)) return
        fs.mkdirs(dest)
        for (entry in fs.list(src)) {
            val target = fs.join(dest, entry.name)
            if (entry.isDirectory) copyTree(entry.path, target) else fs.writeBytes(target, fs.readBytes(entry.path))
        }
    }

    private fun parentOf(path: String): String {
        val idx = path.lastIndexOf('/')
        return if (idx <= 0) "" else path.substring(0, idx)
    }

    private fun relativize(root: String, path: String): String {
        val trimmedRoot = if (root.endsWith('/')) root else "$root/"
        return if (path.startsWith(trimmedRoot)) path.substring(trimmedRoot.length) else path
    }

    companion object {
        const val MAX_READ_BYTES = 1024 * 1024

        private fun seedReadme(title: String): String = """
            |# $title
            |
            |Workspace ini **tidak punya kuota**: simpan file sebesar apa pun, buat session sebanyak apa pun.
            |
            |- Agent punya tools: read_file, write_file, edit_file, list_files, delete_file, exec
            |- Setiap perubahan dicatat di `oplog.jsonl` dan bisa disinkronkan ke cloud
            |
            |Coba minta agent: "buatkan script python yang menghitung fibonacci lalu jalankan".
        """.trimMargin()
    }
}

/** Guard for callers that pass a raw path from the UI instead of a relative one. */
fun WorkspaceEngine.safeResolve(sessionId: String, relative: String): String = try {
    resolve(sessionId, relative)
} catch (e: PathEscapeException) {
    throw WorkspaceFileException(e.message ?: "path tidak valid", "EPATHOUTSIDE")
}

fun WorkspaceEngine.writeChecked(sessionId: String, relative: String, content: String, source: String = "api"): WorkspaceEngine.WriteResult =
    try {
        writeFile(sessionId, relative, content, source)
    } catch (e: PathEscapeException) {
        throw WorkspaceFileException(e.message ?: "path tidak valid", "EPATHOUTSIDE")
    }

/** Convenience for logging structured payloads into the oplog. */
fun WorkspaceEngine.logStructured(sessionId: String, type: String, detail: Map<String, Any?>, source: String = "system") {
    appendOp(sessionId, OpLogEntry(0, type, null, source, nowIso(), detail.mapValues { toJsonValue(it.value) }))
}
